jlm/rvsdg/NodeNormalization.hpp

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_RVSDG_NODENORMALIZATION_HPP
+#define JLM_RVSDG_NODENORMALIZATION_HPP
+
+#include <jlm/rvsdg/node.hpp>
+#include <jlm/rvsdg/region.hpp>
+#include <jlm/util/common.hpp>
+
+#include <functional>
+#include <optional>
+#include <vector>
+
+namespace jlm::rvsdg
+{
+
+class output;
+
+template<class TOperation>
+using NodeNormalization = std::function<
+    std::optional<std::vector<output *>>(const TOperation &, const std::vector<output *> &)>;
+
+template<class TOperation>
+using NodeNormalizationSequenceType = std::function<NodeNormalization<TOperation>(
+    const std::vector<NodeNormalization<TOperation>> &)>;
+
+template<class TOperation>
+NodeNormalizationSequenceType<TOperation> NodeNormalizationSequence =
+    [](const std::vector<NodeNormalization<TOperation>> & nodeNormalizations)
+    -> NodeNormalization<TOperation>
+{
+  return [&](const TOperation & operation,
```

Comment:
The user of this function has to be really careful that the vector still exists when the returned lambda is called.

Reply:
It might be better to use type `std::vector<NodeNormalization<TOperation>>` instead of `const std::vector<NodeNormalization<TOperation>>&` here. Then we would create a copy of the normalizations and everything should be fine.

Reply:
This would also help with your comment above, where the vector should be static at the line above.

Reply:
Alternatively, the lambda can also be [nodeNormalizations]`

Reply:
I would go with the last suggestion. Taking the vector by copy but still capturing it by reference in the lambda would not work.